File: src/stdlib/shepherd.ts
/**
 * `shepherd`: the world outside the program: arguments, environment, input,
 * and running other programs.
 *
 * Security note. `shepherd.run` never uses a shell. It takes a program name and
 * an explicit array of arguments and passes them straight to the operating
 * system, so there is no string that could be re-interpreted as shell syntax
 * and nothing to quote-escape. A program that wants a shell has to ask for one
 * by name (`shepherd.run("bash", ["-c", ...])`), which makes the decision
 * visible in the source and in review. See SECURITY.md.
 */

import { spawnSync } from "node:child_process";
import { readSync } from "node:fs";
import process from "node:process";

import { BaaError } from "../diagnostics/diagnostic.ts";
import type { RuntimeHost } from "../runtime/host.ts";
import { BaaArray, BaaMap } from "../runtime/values.ts";
import type { MapKey, Value } from "../runtime/values.ts";
import { argArray, argString, defineModule, fn, mapOf } from "./define.ts";

export function createShepherd(host: RuntimeHost) {
  return defineModule("shepherd", {
    PLATFORM: process.platform,
    ARCH: process.arch,

    args: fn(0, 0, "Arguments passed to the Baa program after `--`.", () =>
      new BaaArray([...host.argv()]),
    ),

    env: fn(1, 2, "An environment variable, or a fallback when it is unset.", (args, ctx) => {
      const name = argString("shepherd.env", args, 0, ctx.span);
      const value = host.envVar(name);
      return value ?? (args.length > 1 ? args[1]! : null);
    }),

    env_all: fn(0, 0, "Every environment variable as a map.", () => {
      const entries = new Map<MapKey, Value>();
      for (const [key, value] of Object.entries(host.envVars())) entries.set(key, value);
      return new BaaMap(entries);
    }),

    write: fn(1, Number.MAX_SAFE_INTEGER, "Write to stdout without a trailing newline.", (args) => {
      host.write(args.map((value) => (typeof value === "string" ? value : String(value))).join(""));
      return null;
    }),

    write_error: fn(1, Number.MAX_SAFE_INTEGER, "Write to stderr without a trailing newline.", (args) => {
      host.writeError(
        args.map((value) => (typeof value === "string" ? value : String(value))).join(""),
      );
      return null;
    }),

    input: fn(0, 1, "Read one line from stdin, or nil at end of input.", (args, ctx) => {
      if (args.length > 0) host.write(argString("shepherd.input", args, 0, ctx.span));
      return readLine();
    }),

    read_all: fn(0, 0, "Read all of stdin as a single string.", () => {
      const chunks: string[] = [];
      for (;;) {
        const line = readLine();
        if (line === null) break;
        chunks.push(line);
      }
      return chunks.join("\n");
    }),

    run: fn(
      1,
      3,
      "Run a program with an explicit argument array. Never uses a shell.",
      (args, ctx) => {
        const program = argString("shepherd.run", args, 0, ctx.span);
        const programArgs =
          args.length > 1
            ? argArray("shepherd.run", args, 1, ctx.span).items.map((item, index) => {
                if (typeof item !== "string") {
                  throw BaaError.of(
                    "BAA311",
                    ["shepherd.run", "an array of strings", "2", `a non-string at index ${index}`],
                    { span: ctx.span, note: "arguments must all be strings" },
                  );
                }
                return item;
              })
            : [];
        const options = args.length > 2 ? args[2] : null;
        const cwd =
          options instanceof BaaMap && typeof options.entries.get("cwd") === "string"
            ? (options.entries.get("cwd") as string)
            : host.cwd();
        const input =
          options instanceof BaaMap && typeof options.entries.get("input") === "string"
            ? (options.entries.get("input") as string)
            : undefined;

        const result = spawnSync(program, programArgs, {
          cwd,
          encoding: "utf8",
          shell: false,
          ...(input === undefined ? {} : { input }),
        });
        if (result.error !== undefined) {
          throw BaaError.of("BAA301", [`could not run \`${program}\`: ${result.error.message}`], {
            span: ctx.span,
            note: "the program did not start",
            help: "Check that it is installed and on PATH.",
          });
        }
        return mapOf({
          code: result.status ?? -1,
          out: result.stdout ?? "",
          err: result.stderr ?? "",
        });
      },
    ),

    exit: fn(0, 1, "Exit with a status code (default 0).", (args, ctx) => {
      const code = args.length > 0 ? args[0] : 0;
      if (typeof code !== "number" || !Number.isInteger(code)) {
        throw BaaError.of("BAA311", ["shepherd.exit", "a whole number", "1", typeof code], {
          span: ctx.span,
        });
      }
      host.exit(code);
      return null;
    }),
  });
}

/**
 * Synchronous line read from stdin. Baa's interpreter is synchronous, so this
 * blocks by design. Returns `null` at end of input.
 */
function readLine(): string | null {
  const buffer = Buffer.alloc(1);
  let line = "";
  for (;;) {
    let read = 0;
    try {
      read = readSync(0, buffer, 0, 1, null);
    } catch (error) {
      const code = (error as NodeJS.ErrnoException).code;
      if (code === "EAGAIN") continue;
      if (code === "EOF") break;
      throw error;
    }
    if (read === 0) break;
    const ch = buffer.toString("utf8");
    if (ch === "\n") return line.endsWith("\r") ? line.slice(0, -1) : line;
    line += ch;
  }
  return line.length > 0 ? line : null;
}

File: rust/crates/baa-native/src/stdlib/shepherd.rs
//! `shepherd`: the world outside the program. A port of
//! `src/stdlib/shepherd.ts`.
//!
//! Security note, carried over unchanged: `shepherd.run` never uses a shell.
//! It takes a program name and an explicit array of arguments and hands them
//! to the operating system, so there is no string that could be re-read as
//! shell syntax and nothing to quote-escape. A program that wants a shell has
//! to ask for one by name, which puts the decision in the source and in
//! review. See SECURITY.md.
//!
//! `write` goes to the same stream `baa` writes to, not to `print!`. In a
//! windowed application that stream is a sink, so a program that prints has no
//! console window flashing behind it and no hidden file nobody reads.

use std::io::{Read, Write};
use std::process::{Command, Stdio};
use std::rc::Rc;

use crate::ast::Span;
use crate::interp::{Flow, Interpreter, Res};
use crate::value::{display, BaaMap, MapKey, Module, Native, Value};

use super::{map_value, module_of, need_array, need_string, option_str};

const FUNCTIONS: &[(&str, usize, usize)] = &[
    ("args", 0, 0),
    ("env", 1, 2),
    ("env_all", 0, 0),
    ("write", 1, usize::MAX),
    ("write_error", 1, usize::MAX),
    ("input", 0, 1),
    ("read_all", 0, 0),
    ("run", 1, 3),
    ("exit", 0, 1),
];

/// The names Node reports, because the reference implementation reports Node's
/// and a program that branches on the platform has to see the same word from
/// both runtimes.
const PLATFORM: &str = if cfg!(target_os = "windows") {
    "win32"
} else if cfg!(target_os = "macos") {
    "darwin"
} else {
    std::env::consts::OS
};

const ARCH: &str = if cfg!(target_arch = "x86_64") {
    "x64"
} else if cfg!(target_arch = "aarch64") {
    "arm64"
} else if cfg!(target_arch = "x86") {
    "ia32"
} else {
    std::env::consts::ARCH
};

pub fn module() -> Rc<Module> {
    let mut exports: Vec<(&str, Value)> =
        vec![("PLATFORM", Value::str(PLATFORM)), ("ARCH", Value::str(ARCH))];
    for (name, min, max) in FUNCTIONS {
        exports.push((
            name,
            Value::Native(Rc::new(Native {
                name: format!("shepherd.{name}"),
                method: name,
                min_args: *min,
                max_args: *max,
                call,
                receiver: None,
            })),
        ));
    }
    module_of("shepherd", exports)
}

fn call(interp: &mut Interpreter, native: &Native, args: Vec<Value>, span: Span) -> Res<Value> {
    let name = native.name.clone();
    Ok(match native.method {
        "args" => Value::array(interp.argv.iter().map(Value::str).collect()),

        "env" => match std::env::var(&*need_string(interp, &name, &args, 0, span)?) {
            Ok(value) => Value::str(value),
            Err(_) => args.get(1).cloned().unwrap_or(Value::Nil),
        },

        "env_all" => {
            let mut map = BaaMap::new();
            for (key, value) in std::env::vars() {
                map.set(MapKey::Str(Rc::from(key.as_str())), Value::str(value));
            }
            Value::map(map)
        }

        "write" => {
            let _ = interp.out.write_all(joined(&args).as_bytes());
            Value::Nil
        }

        "write_error" => {
            let _ = std::io::stderr().write_all(joined(&args).as_bytes());
            Value::Nil
        }

        "input" => {
            if let Some(prompt) = args.first() {
                let text = need_string(interp, &name, &args, 0, span)?;
                let _ = prompt;
                let _ = interp.out.write_all(text.as_bytes());
                let _ = interp.out.flush();
            }
            match read_line() {
                Some(line) => Value::str(line),
                None => Value::Nil,
            }
        }

        "read_all" => {
            // The reference reads line by line and joins with "\n", which
            // drops a trailing newline. Matched deliberately: differing here
            // would show up as one invisible character.
            let mut lines: Vec<String> = Vec::new();
            while let Some(line) = read_line() {
                lines.push(line);
            }
            Value::str(lines.join("\n"))
        }

        "run" => {
            let program = need_string(interp, &name, &args, 0, span)?;
            let mut arguments: Vec<String> = Vec::new();
            if args.len() > 1 {
                for (index, item) in need_array(interp, &name, &args, 1, span)?.iter().enumerate() {
                    match item {
                        Value::Str(text) => arguments.push(text.to_string()),
                        _ => {
                            return Err(Flow::Err(
                                interp
                                    .error(
                                        "BAA311",
                                        vec![
                                            name.to_string(),
                                            "an array of strings".into(),
                                            "2".into(),
                                            format!("a non-string at index {index}"),
                                        ],
                                        span,
                                    )
                                    .with_note("arguments must all be strings"),
                            ))
                        }
                    }
                }
            }
            let options = args.get(2);
            let input = option_str(options, "input");

            let mut command = Command::new(&*program);
            command.args(&arguments);
            if let Some(cwd) = option_str(options, "cwd") {
                command.current_dir(&*cwd);
            }
            command
                .stdin(if input.is_some() { Stdio::piped() } else { Stdio::null() })
                .stdout(Stdio::piped())
                .stderr(Stdio::piped());

            let mut child = match command.spawn() {
                Ok(child) => child,
                Err(error) => {
                    return Err(Flow::Err(
                        interp
                            .error(
                                "BAA301",
                                vec![format!("could not run `{program}`: {error}")],
                                span,
                            )
                            .with_note("the program did not start")
                            .with_help("Check that it is installed and on PATH."),
                    ))
                }
            };
            if let (Some(text), Some(mut pipe)) = (input, child.stdin.take()) {
                let _ = pipe.write_all(text.as_bytes());
            }
            match child.wait_with_output() {
                Ok(output) => map_value(vec![
                    ("code", Value::Number(output.status.code().unwrap_or(-1) as f64)),
                    ("out", Value::str(String::from_utf8_lossy(&output.stdout))),
                    ("err", Value::str(String::from_utf8_lossy(&output.stderr))),
                ]),
                Err(error) => {
                    return Err(Flow::Err(
                        interp
                            .error("BAA301", vec![format!("`{program}` failed: {error}")], span)
                            .with_note("the program started but could not be waited for"),
                    ))
                }
            }
        }

        "exit" => {
            let code = match args.first() {
                None | Some(Value::Nil) => 0.0,
                Some(Value::Number(value)) if value.fract() == 0.0 => *value,
                Some(other) => {
                    return Err(Flow::Err(interp.error(
                        "BAA311",
                        vec![name.to_string(), "a whole number".into(), "1".into(), other.describe()],
                        span,
                    )))
                }
            };
            return Err(Flow::Exit(code as i32));
        }

        _ => Value::Nil,
    })
}

/// `write` and `write_error` take any number of values and concatenate them,
/// with strings used as they are and everything else displayed.
fn joined(args: &[Value]) -> String {
    args.iter()
        .map(|value| match value {
            Value::Str(text) => text.to_string(),
            other => display(other),
        })
        .collect()
}

/// One line from standard input, without its terminator, or `None` at the end
/// of input. Blocking by design: the interpreter is synchronous.
fn read_line() -> Option<String> {
    let mut line = String::new();
    let mut byte = [0u8; 1];
    let mut stdin = std::io::stdin();
    loop {
        match stdin.read(&mut byte) {
            Ok(0) => break,
            Ok(_) => {
                if byte[0] == b'\n' {
                    if line.ends_with('\r') {
                        line.pop();
                    }
                    return Some(line);
                }
                line.push(byte[0] as char);
            }
            Err(_) => break,
        }
    }
    if line.is_empty() {
        None
    } else {
        Some(line)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_reports_the_platform_names_node_reports() {
        // A program that branches on `shepherd.PLATFORM` has to see the same
        // word from both runtimes, and Node's words are not Rust's.
        assert!(["win32", "darwin", "linux"].contains(&PLATFORM), "{PLATFORM}");
        assert!(["x64", "arm64", "ia32"].contains(&ARCH), "{ARCH}");
    }

    #[test]
    fn joining_uses_strings_as_they_are() {
        assert_eq!(joined(&[Value::str("a"), Value::Number(1.0), Value::Nil]), "a1nil");
    }
}
